### PR TITLE
WIP - Read parquet files

### PR DESCRIPTION
**Pull request by Moohan**
This is the experimental code for reading a parquet version of the SLFs. Changes include.

- New functions `read_slf_*_parquet` which read a parquet file from `sourcedev`.
- A `dev` parameter to allow using files from `sourcedev`.

I think the `read_slf_*_parquet` mirrors the functionality of `read_slf_*` but testing would be appreciated.

### Still to do
- [x] Make the parquet functions the default / only ones.
- [x] Check and update the tests so they work with the new files.
- [ ] Add tests for the `dev` parameter.
- [ ] Ensure there are parquet files available for all years.
- [x] Update `dev` parameter to gracefully error when a file doesn't exist.
- [x] Update `dev` parameter to give messaging so that it's not used by accident.
- [x] Improve the documentation of the new parquet functions.

### CODE REVIEW

File: R/read_slf_parquet.R [hunk 1]
```diff
@@ -63,11 +63,11 @@ read_slf_parquet <- function(
       if (!is.null(partnerships)) {
         slf_table <- dplyr::filter(
           slf_table,
-          .data$partnership %in% partnerships
+          .data$hscp2019 %in% partnerships
         )
       }
       if (remove_partnership_var) {
-        slf_table <- dplyr::select(slf_table, -"hscp2018")
+        slf_table <- dplyr::select(slf_table, -"hscp2019")
```
Comment (Moohan) [inline R/read_slf_parquet.R]: I think this should all be hscp2018 (the partnership bit was a mistake though!)

Comment (shintoLamp) [reply R/read_slf_parquet.R]: So in the parquet file I was testing on, we've got the option of `hscp` or `hscp2019`, `hscp2018` doesn't exist. Maybe something to consider for the SLFs in general

Comment (Moohan) [reply R/read_slf_parquet.R]: Yep that's an issue! I'll see if I can find out what's happened in the SLFs.

The `partnerships` lookup uses the hscp2018 codes, so it definitely needs to be compared against the variable containing those codes though.

**Comment by Moohan** [comment]
@shintoLamp Do you have time to check this again please?

**Comment by Jennit07** [comment]
I think the changes are working for reading in a file but i ran into this error when testing what happens when i select `anon_chi`. I think this is also defaulted to set `dev` as TRUE is this correct or should the default be FALSE when trying to read from hscdiip?

> ep_1718 <- read_slf_episode("1718", col_select = c("anon_chi", "recid"))

ℹ "anon_chi" has been replaced with "chi" in `col_select` as you requested the development file.
Error in `purrr::map()` at slfhelper/R/read_slf.R:80:2:
ℹ In index: 1.
Caused by error in `arrow::read_parquet()` at slfhelper/R/read_slf.R:83:6:
! Can't subset columns that don't exist.
✖ Column `chi` doesn't exist.
Run `rlang::last_trace()` to see where the error occurred.

File: R/read_slf.R [hunk 2]
```diff
@@ -3,185 +3,212 @@
...
-        partnerships = unique(partnerships),
-        columns = unique(columns),
-        ...
-      )
+read_slf_individual <- function(
+    year,
+    col_select = NULL,
+    partnerships = NULL,
+    as_data_frame = TRUE,
+    dev = TRUE,
```
Comment (Jennit07) [inline R/read_slf.R]:
```suggestion
    dev = FALSE,
```
Should default be false here?

File: R/read_slf.R [hunk 3]
```diff
@@ -3,185 +3,212 @@
...
-        columns = unique(columns),
-        ...
-      )
+read_slf_episode <- function(
+    year,
+    col_select = NULL,
+    partnerships = NULL,
+    recids = NULL,
+    as_data_frame = TRUE,
+    dev = TRUE,
```
Comment (Jennit07) [inline R/read_slf.R]:
```suggestion
    dev = FALSE,
```
Should dev be false here?

**Comment by Moohan** [comment]
> I think the changes are working for reading in a file but i ran into this error when testing what happens when i select `anon_chi`. I think this is also defaulted to set `dev` as TRUE is this correct or should the default be FALSE when trying to read from hscdiip?
> 
> > ep_1718 <- read_slf_episode("1718", col_select = c("anon_chi", "recid"))
> 
> ℹ "anon_chi" has been replaced with "chi" in `col_select` as you requested the development file. Error in `purrr::map()` at slfhelper/R/read_slf.R:80:2: ℹ In index: 1. Caused by error in `arrow::read_parquet()` at slfhelper/R/read_slf.R:83:6: ! Can't subset columns that don't exist. ✖ Column `chi` doesn't exist. Run `rlang::last_trace()` to see where the error occurred.

Is this working now?

You were right about the `dev` param that was just a mistake, but actually, the code that's swapping out `anon_chi` for `chi` isn't needed anymore either!

**Comment by Jennit07** [comment]
> > I think the changes are working for reading in a file but i ran into this error when testing what happens when i select `anon_chi`. I think this is also defaulted to set `dev` as TRUE is this correct or should the default be FALSE when trying to read from hscdiip?
> > > ep_1718 <- read_slf_episode("1718", col_select = c("anon_chi", "recid"))
> > 
> > 
> > ℹ "anon_chi" has been replaced with "chi" in `col_select` as you requested the development file. Error in `purrr::map()` at slfhelper/R/read_slf.R:80:2: ℹ In index: 1. Caused by error in `arrow::read_parquet()` at slfhelper/R/read_slf.R:83:6: ! Can't subset columns that don't exist. ✖ Column `chi` doesn't exist. Run `rlang::last_trace()` to see where the error occurred.
> 
> Is this working now?
> 
> You were right about the `dev` param that was just a mistake, but actually, the code that's swapping out `anon_chi` for `chi` isn't needed anymore either!

Yeah, i think this is working correctly now. For the tests i think we need to rename the files in hscdiip to remove the `20-` as this was causing them to fail

**Comment by Jennit07** [comment]
`tests-file-exists` will still be failing in the tests as the new parquet files have not transferred into hscdiip yet. I have updated the tests so that they will pass when these have been copied over. This was tested by using 1718 as an example which is already copied into hscdiip

**Comment by Jennit07** [comment]
I need to look at this again on monday. When implementing `dplyr::slice_sample` this was not resolving the tests

**Comment by Jennit07** [comment]
Tests have been updated and should hopefully work as expected now. I have tested this with 1718 parquet file in hscdiip but many tests will still fail until they are transferred over. Once these are transferred over, they should pass

**Pull request merged**